[processor/transform] more readme refactoring (#37909)
open-telemetry/opentelemetry-collector-contrib: processor/transformprocessor/README.md

<!-- end autogenerated section -->

> [!NOTE]
> This documentation applies only to version `0.120.0` and later. **Configuration from previous version is still supported**, but no longer documented in this README. For information on earlier versions, please refer to the previous [documentation](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/release/0.119.x/processor/transformprocessor/README.md).

The Transform Processor modifies telemetry based on configuration using the [OpenTelemetry Transformation Language](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/pkg/ottl) (OTTL).

For each signal type, the processor takes a list of statements and executes them against the incoming telemetry, following the order specified in the configuration.
Each statement can access and transform telemetry using functions, and allows the use of a condition to help decide whether the function should be executed.

> - [Writing OTTL](#writing-ottl)

Author: The more I read through this doc the more important it seems to learn how to write OTTL before even worrying about configuring the processor. Switched this and the `Config` section to help new users find `Writing OTTL` first.

Reviewer: Could we link out to a section on writing OTTL inside the OTTL docs instead? I'd like for the transform processor readme to cut to the chase and show how it's configured, and I think the callouts should be enough of an attention getter to help new users find out how to write OTTL.

Author: Should I move the entire `Writing OTTL` section to `pkg/ottl/README`?

Reviewer: That's what I was thinking. It would also group things a bit better in my opinion, since all OTTL-specific documentation would be in the OTTL readme.

Author: Ok I'll make that change.

Author: Updated

For each signal type, the processor takes a list of statements and executes them against the incoming telemetry, following the order specified in the configuration.
Each statement can access and transform telemetry using functions, and allows the use of a condition to help decide whether the function should be executed.

- [Config](#config)
- [Grammar](#grammar)
- [Supported functions](#supported-functions)
- [Examples](#examples)
- [Troubleshooting](#troubleshooting)
- [Contributing](#contributing)
- [Feature Gate](#feature-gate)

## Config

### General Config

> [!NOTE]
> If you don't know how to write OTTL statements yet, first see OTTL's [Getting Started](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/pkg/ottl/README.md#getting-started) docs.

```yaml
transform:
  error_mode: ignore
  <trace|metric|log>_statements: []
```

The Transform Processor's primary configuration section is broken down by signal (traces, metrics, and logs)
and allows you to configure a list of statements for the processor to execute. The list can be made of:

- OTTL statements. This option will meet most user's needs. See [Basic Config](#basic-config) for more details.

> - Objects, which allows users to apply configuration options to a specific-list of statements. See [Advanced Config](#advanced-config) for more details.

Reviewer: ```suggestion
- Objects, which allows users to apply configuration options to a specific list of statements. See [Advanced Config](#advanced-config) for more details.
```

| error_mode | description                                                                                                                                 |
|------------|---------------------------------------------------------------------------------------------------------------------------------------------|
| ignore     | The processor ignores errors returned by statements, logs the error, and continues on to the next statement.  This is the recommended mode. |
| silent     | The processor ignores errors returned by statements, does not log the error, and continues on to the next statement.                        |
| propagate  | The processor returns the error up the pipeline.  This will result in the payload being dropped from the collector.                         |

### Basic Config

> [!NOTE]
> If you don't know how to write OTTL statements yet, first see OTTL's [Getting Started](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/pkg/ottl/README.md#getting-started) docs.

The basic configuration style allows you to configure OTTL statements as a list,
without worrying about extra configurations.

> **This is the simplest way to configure the Transform Processor.** If you need global conditions or specific error modes see [Statement Group Config](#statement-group-config).

Reviewer 2: ```suggestion
**This is the simplest way to configure the Transform Processor.** If you need global conditions or specific error modes see [Advanced Config](#advanced-config).
```

<!-- end autogenerated section -->

> [!NOTE]
> This documentation applies only to version `0.120.0` and later. **Configuration from previous version is still supported**, but no longer documented in this README. For information on earlier versions, please refer to the previous [documentation](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/release/0.119.x/processor/transformprocessor/README.md).

The Transform Processor modifies telemetry based on configuration using the [OpenTelemetry Transformation Language](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/pkg/ottl) (OTTL).

> For each signal type, the processor takes a list of statements and executes statements against the incoming telemetry in the order specified in the config.

Reviewer 2: ```suggestion
For each signal type, the processor takes a list of statements and executes them against the incoming telemetry, following the order specified in the configuration.
```

> Each statement can access and transform telemetry using functions and allow the use of a condition to help decide whether the function should be executed.

Reviewer: ```suggestion
Each statement can access and transform telemetry using functions, and allows the use of a condition to help decide whether the function should be executed.
```
I know this sentence wasn't changed, but as-is it's a bit hard to parse. I'd also be okay just removing this and we can explain these concepts inside docs for OTTL.

> Each statement can access and transform telemetry using functions and allow the use of a condition to help decide whether the function should be executed.

Reviewer 2: ```suggestion
Each statement can access and transform telemetry using functions, and allow the use of a condition to help decide whether the function should be executed.
```

- [Config](#config)
- [Writing OTTL](#writing-ottl)
- [Grammar](#grammar)
- [Supported functions](#supported-functions)
- [Examples](#examples)
- [Troubleshooting](#troubleshooting)
- [Contributing](#contributing)
- [Feature Gate](#feature-gate)

## Config

The Transform Processor allows configuring OTTL statements for traces, metrics, and logs.

> > [!NOTE]

Reviewer 2: This note is equals and visually very close to the below note, should we remove this one?

Author: I can't see which note you're referring to anymore, but I would like to keep it at the top of the 2 config styles.

Author: I'd like to keep this note as a warning to anyone who read through `Writing OTTL` that this is something they can skip unless they are interested.

> > [!NOTE]
> > If you don't know how to write OTTL statements yet, first see [Writing OTTL](#writing-ottl).
> 

Reviewer: ```suggestion
```
We can probably remove this note here, we also have it right below.

### General Config

> [!NOTE]
> If you don't know how to write OTTL statements yet, first see [Writing OTTL](#writing-ottl).

```yaml
transform:
  error_mode: ignore
  <trace|metric|log>_statements: []
```

Only certain OTTL Path prefixes can be used for each signal:

| Signal            | Path Prefix Values                             |
|-------------------|------------------------------------------------|
| trace_statements  | `resource`, `scope`, `span`, and `spanevent`   |
| metric_statements | `resource`, `scope`, `metric`, and `datapoint` |
| log_statements    | `resource`, `scope`, and `log`                 |

This means, for example, that you cannot use the Path `span.attributes` within the `log_statements` configuration section.

> `error_mode`: determines how the processor reacts to errors that occur while processing a statement.

Reviewer: ```suggestion
`error_mode`: determines how the processor treats errors that occur while processing a statement.
```
Just a personal preference.

If the top-level `error_mode` is not specified, `propagate` will be used.

> The top-level `error_mode` can be overridden at context statements level, offering more granular control over error handling. If context statement `error_mode` is not used, the top-level `error_mode` is used.

Reviewer 2: ```suggestion
The top-level `error_mode` can be overridden at statement group level, offering more granular control over error handling. If the statement group `error_mode` is not specified, the top-level `error_mode` is applied.
```
Maybe, to keep the naming similar to the new config name `Statement Group Config`?

| error_mode | description                                                                                                                                 |
|------------|---------------------------------------------------------------------------------------------------------------------------------------------|
| ignore     | The processor ignores errors returned by statements, logs the error, and continues on to the next statement.  This is the recommended mode. |
| silent     | The processor ignores errors returned by statements, does not log the error, and continues on to the next statement.                        |
| propagate  | The processor returns the error up the pipeline.  This will result in the payload being dropped from the collector.                         |

> ### Flat Config

Reviewer: Would we be able to get away without introducing any new concepts and just say that each item in the `X_statements` list is:

1. "Basic config": A statement
2. "Advanced config": A nested config object within the list that contains a list of statements and options for configuring their execution.

Author: Ya those terms work for me, I'll make the updates.

Author: updated

Format:

```yaml
transform:
  error_mode: ignore
  <trace|metric|log>_statements:
    - string
    - string
    - string
```

Example: 

```yaml
transform:
  error_mode: ignore
  trace_statements:
    - keep_keys(span.attributes, ["service.name", "service.namespace", "cloud.region", "process.command_line"])
    - replace_pattern(span.attributes["process.command_line"], "password\\=[^\\s]*(\\s?)", "password=***")
    - limit(span.attributes, 100, [])
    - truncate_all(span.attributes, 4096)
  metric_statements:
    - keep_keys(resource.attributes, ["host.name"])
    - truncate_all(resource.attributes, 4096)
    - set(metric.description, "Sum") where metric.type == "Sum"
    - convert_sum_to_gauge() where metric.name == "system.processes.count"
    - convert_gauge_to_sum("cumulative", false) where metric.name == "prometheus_metric"
  log_statements:
    - set(log.severity_text, "FAIL") where log.body == "request failed"
    - replace_all_matches(log.attributes, "/user/*/list/*", "/user/{userId}/list/{listId}")
    - replace_all_patterns(log.attributes, "value", "/account/\\d{4}", "/account/{accountId}")
    - set(log.body, log.attributes["http.route"])
```

> If you're interested in how OTTL parses these statements, see [Conetxt Inference](#context-inference).

Reviewer: ```suggestion
If you're interested in how OTTL parses these statements, see [Context Inference](#context-inference).
```

### Statement Group Config

> [!NOTE]
> If you don't know how to write OTTL statements yet, first see [Writing OTTL](#writing-ottl).

For more complex use cases you may need to use Statement Groups to group related OTTL statements.

```yaml

>   - error_mode: propagate

Reviewer: Can you nest this in an `X_statements` object to make it more obvious how it fits into the overall config?

Author: updated

        - set(log.body, log.attributes["http.route"])
```

Like the flat configuration style, the OTTL Context in which the statement in the group are executed

> is [inferred](#context-inference), but this time from ALL the statement.

Reviewer: ```suggestion
is [inferred](#context-inference), but this time from ALL the statements.
```

> In some situations the automatic Context inference is not possible. For example:

Reviewer: Could we try to simplify this to not mention contexts? We can still point to the Context inference section at the end and say to consult that for more information.

If possible, I think it would hopefully be sufficient to describe that data in the Collector is nested in lists, and while you can't reach into more deeply-nested items, you can reach into less deeply-nested items. However, the `statements` sections each all need to run at the same level of nesting, so cases like this where you're working with two types of data that are at different levels of nesting can cause conflicts. Describing the OTLP hierarchy in a separate section in this readme may help illustrate how the data types relate to one another.

It's also possible we'd just be dancing around the concept of contexts too much here, so we may just have to go into that detail. But if possible it would be nice to hide that concept as much as possible and just have users focus on whatever data type they want to work with.

Author: I'll try to reword it to hide the `context` concept

Author: updated

```yaml
metric_statements:
  - statements:
    - convert_sum_to_gauge() where metric.name == "system.processes.count"
    - limit(datapoint.attributes, 100, ["host.name"])
```

In this configuration, while the `datapoint` context appears suitable for parsing both `metric`
and `datapoint` data, it does not fulfill the function requirement since `convert_sum_to_gauge` is
a specific `metric` function. Conversely, the `metric` context cannot parse `datapoint`, as it
cannot access "lower" items in the protobuf definition. Therefore, the conflicting statements must
be placed into separate statement groups:

```yaml
metric_statements:
  - statements:
    - limit(datapoint.attributes, 100, ["host.name"])
  - statements:
    - convert_sum_to_gauge() where metric.name == "system.processes.count" 
```

> Alternatively, for simplicity, you can use the [flat configuration](#context-inferred-configurations) style:

Reviewer 2: ```suggestion
Alternatively, for simplicity, you can use the [flat configuration](#flat-config) style:
```

```yaml
metric_statements:
  - limit(datapoint.attributes, 100, ["host.name"])
  - convert_sum_to_gauge() where metric.name == "system.processes.count" 
```

## Writing OTTL

The Transform Processor utilizes OTTL statements to transform telemetry. An OTTL statement is made up of 2 parts:
  1. A function that transforms telemetry

>   2. Optionally a condition that determines whether the function is executed.

Reviewer: ```suggestion
  2. Optionally, a condition that determines whether the function is executed.
```

    - limit(datapoint.attributes, 100, ["host.name"])
```

In this configuration, the `datapoint` Path prefixed is used in the same group of statements as the `convert_sum_to_gauge`

> function. Since `convert_sum_to_gauge` can only be used with the Metric signal, not the Datapoint signal, but the list

Reviewer: I feel like calling data points a "signal" is playing a bit fast and loose with the term, could we just say "metrics" and "data points"?

Author: Ya
